File: src/main/java/core/procedures/strings/IndexOf.java
package core.procedures.strings;

import core.exceptions.WrongTypeException;
import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.utils.Utils;

public final class IndexOf extends AFn {

  public IndexOf() {
    super(new FnArgsBuilder().min(2).max(3).mandatory(new Class[]{CharSequence.class}).build());
  }

  @Override
  public String getName() {
    return "index-of";
  }

  @Override
  public Integer apply(Object... args) {
    if (!(args[1] instanceof CharSequence) && !(args[1] instanceof Character)) {
      throw new WrongTypeException(getName(), "String or Character", args[1]);
    }
    if (args.length == 3) {
      Object index = args[2];
      if (!Utils.INSTANCE.isReal(index)) {
        throw new WrongTypeException(getName(), "Real", index);
      }
      if (args[1] instanceof Character) {
        return args[0].toString().indexOf((char)args[1], ((Number)index).intValue());
      }
      return args[0].toString().indexOf(args[1].toString(), ((Number)index).intValue());
    }
    if (args[1] instanceof Character) {
      return args[0].toString().indexOf((char)args[1]);
    }
    return args[0].toString().indexOf(args[1].toString());
  }
}

File: src/main/java/core/procedures/strings/IndexOf.kt
package core.procedures.strings

import core.exceptions.WrongTypeException
import core.procedures.AFn
import core.procedures.Arity.Range
import core.scm.Type

class IndexOf : AFn<Any?, Int>(name = "index-of", isPure = true, arity = Range(2, 3),
                               mandatoryArgsTypes = arrayOf(CharSequence::class.java)) {

    override operator fun invoke(args: Array<out Any?>) = when {
        args[1] !is CharSequence && args[1] !is Char -> throw WrongTypeException(name, "String or Character", args[1])
        args.size == 3 -> {
            val index = args[2]
            Type.assertType(name, index, Type.Real::class.java)
            when (args[1]) {
                is Char -> args[0].toString().indexOf(args[1] as Char, (index as Number).toInt())
                else    -> args[0].toString().indexOf(args[1].toString(), (index as Number).toInt())
            }
        }
        args[1] is Char -> args[0].toString().indexOf((args[1] as Char))
        else -> args[0].toString().indexOf(args[1].toString())
    }
}
